ext/opcache/Optimizer/block_pass.c

```diff
@@ -466,6 +466,19 @@ static void zend_optimize_block(zend_basic_block *block, zend_op_array *op_array
 								MAKE_NOP(opline);
 								++(*opt_count);
 								break;
+							case ZEND_TYPE_CHECK:
+								if (opline->opcode == ZEND_BOOL_NOT) {
+									if (src->extended_value == MAY_BE_RESOURCE || src->extended_value == (MAY_BE_ANY - MAY_BE_RESOURCE)) {
+										/* is_resource() is a special case - it returns false if the resource is closed. Don't convert to/from it. */
```

Comment:
I don't like the special casing of resources, where only MAY_BE_RESOURCE has a special meaning, but a union including MAY_BE_RESOURCE does not. I think it may make sense to make ZEND_TYPE_CHECK a pure type check and add a separate opcode just for resources.

Reply:
> I don't like the special casing of resources, where only MAY_BE_RESOURCE has a special meaning, but a union including MAY_BE_RESOURCE does not. I think it may make sense to make ZEND_TYPE_CHECK a pure type check and add a separate opcode just for resources.

I'd also thought about that - is there anything that needs to be done to approve/validate new opcodes? (e.g. benchmarks)

Reply:
Also, ZEND_TYPE_CHECK is declared with TMPVAR spec, meaning that it has to check for undefined variables, even for TMP vals (e.g. `return is_string($this->val) === false` for IS_FALSE). Can it be changed to TMP|VAR, if it'll be used more often?